File: subscription/src/main/scala/com/timgroup/eventsubscription/healthcheck/ChaserHealth.scala
package com.timgroup.eventsubscription.healthcheck

import com.timgroup.eventsubscription.ChaserListener
import com.timgroup.eventsubscription.util.Clock
import com.timgroup.tucker.info.Status.{OK, WARNING, CRITICAL}
import com.timgroup.tucker.info.{Component, Report, Status}
import org.joda.time.{DateTime, Seconds}
import org.slf4j.LoggerFactory

class ChaserHealth(name: String, clock: Clock) extends Component("event-store-chaser-" + name, "Eventstore chaser health (" + name + ")") with ChaserListener {
  @volatile private var lastPollTimestamp: Option[DateTime] = None
  @volatile private var currentVersion: Long = 0

  override def getReport = {
    lastPollTimestamp match {
      case None => new Report(WARNING, "Awaiting initial catchup. Current version: " + currentVersion)
      case Some(timestamp) => {
        val seconds = Seconds.secondsBetween(timestamp, clock.now()).getSeconds
        if (seconds > 30) {
          new Report(CRITICAL, "potentially stale. Last up-to-date at at %s. (%s seconds ago).".format(timestamp, seconds))
        } else if (seconds > 5) {
          new Report(WARNING, "potentially stale. Last up-to-date at at %s. (%s seconds ago).".format(timestamp, seconds))
        } else {
          new Report(OK, "up-to-date at at %s. (%s seconds ago). Current version: %s".format(timestamp, seconds, currentVersion))
        }
      }
    }
  }

  override def transientFailure(e: Exception): Unit = {
    LoggerFactory.getLogger(getClass).warn("Failure chasing eventstream.", e)
  }

  override def chaserReceived(version: Long): Unit = {
    currentVersion = version
  }

  override def chaserUpToDate(version: Long): Unit = {
    lastPollTimestamp = Some(clock.now())
  }
}

File: subscription/src/main/java/com/timgroup/eventsubscription/healthcheck/ChaserHealth.java
package com.timgroup.eventsubscription.healthcheck;

import java.time.Clock;
import java.time.Duration;
import java.time.Instant;

import com.timgroup.eventstore.api.Position;
import com.timgroup.eventsubscription.ChaserListener;
import com.timgroup.tucker.info.Component;
import com.timgroup.tucker.info.Report;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static com.timgroup.tucker.info.Status.CRITICAL;
import static com.timgroup.tucker.info.Status.INFO;
import static com.timgroup.tucker.info.Status.OK;
import static com.timgroup.tucker.info.Status.WARNING;
import static java.lang.String.format;

public class ChaserHealth extends Component implements ChaserListener {
    private static final Logger LOG = LoggerFactory.getLogger(ChaserHealth.class);
    private static final Duration MIN_WARNING_THRESHOLD = Duration.ofSeconds(5);
    private static final Duration MIN_CRITICAL_THRESHOLD = Duration.ofSeconds(30);
    private final Clock clock;
    private final Duration warningThreshold;
    private final Duration criticalThreshold;

    private volatile Instant lastPollTimestamp;
    private volatile Position currentPosition;

    public ChaserHealth(String name, Clock clock) {
        this(name, clock, Duration.ofSeconds(1));
    }

    public ChaserHealth(String name, Clock clock, Duration subscriptionRunFrequency) {
        super("event-store-chaser-" + name, "Eventstore chaser health (" + name + ")");
        this.clock = clock;
        this.warningThreshold = max(MIN_WARNING_THRESHOLD, subscriptionRunFrequency.multipliedBy(5));
        this.criticalThreshold = max(MIN_CRITICAL_THRESHOLD, subscriptionRunFrequency.multipliedBy(10));
    }

    @Override
    public Report getReport() {
        Instant timestamp = lastPollTimestamp;
        if (timestamp != null) {
            Duration duration = Duration.between(timestamp, Instant.now(clock));
            if (duration.compareTo(criticalThreshold) > 0) {
                return new Report(CRITICAL, format("potentially stale. Last up-to-date at at %s. (%s ago).", timestamp, duration));
            } else if (duration.compareTo(warningThreshold) > 0) {
                return new Report(WARNING, format("potentially stale. Last up-to-date at at %s. (%s ago).", timestamp, duration));
            } else {
                return new Report(OK, format("up-to-date at at %s. (%s ago). Current version: %s", timestamp, duration, currentPosition));
            }
        }

        return new Report(INFO, "Awaiting initial catchup. Current version: " + currentPosition);
    }

    @Override
    public void transientFailure(Exception e) {
        LOG.warn("Failure chasing eventstream.", e);
    }

    @Override
    public void chaserReceived(Position position) {
        currentPosition = position;
    }

    @Override
    public void chaserUpToDate(Position position) {
        currentPosition = position;
        lastPollTimestamp = Instant.now(clock);
    }

    private static Duration max(Duration d1, Duration d2) {
        return d1.compareTo(d2) > 0 ? d1 : d2;
    }
}
